File: tests/typescript/steps/performance.steps.ts
import { Given, When, Then, Before, After, setDefaultTimeout } from "@cucumber/cucumber";
import {
  Aptos,
  AptosConfig,
  Network,
  Account,
  Ed25519PrivateKey,
  AccountAddress,
} from "@aptos-labs/ts-sdk";
import { sha3_256 } from "@noble/hashes/sha3.js";
import { TestWorld, BenchmarkResult } from "../support/world";

// Set longer timeout for performance tests (10 minutes)
setDefaultTimeout(600000);

interface PerformanceResults {
  sdk: string;
  version: string;
  runtime: string;
  network: string;
  timestamp: string;
  environment: {
    platform: string;
    arch: string;
    nodeVersion: string;
  };
  results: Record<string, number>;
}

// Global performance results collector
const performanceResults: PerformanceResults = {
  sdk: "typescript",
  version: "5.2.0",
  runtime: `bun ${process.versions.bun || "unknown"}`,
  network: "devnet",
  timestamp: new Date().toISOString(),
  environment: {
    platform: process.platform,
    arch: process.arch,
    nodeVersion: process.version,
  },
  results: {},
};

function calculateStats(samples: number[]): BenchmarkResult {
  const sorted = [...samples].sort((a, b) => a - b);
  const sum = sorted.reduce((acc, val) => acc + val, 0);
  const avg = sum / sorted.length;
  const p95Index = Math.floor(sorted.length * 0.95);
  const p95 = sorted[p95Index] || sorted[sorted.length - 1];
  const min = sorted[0];
  const max = sorted[sorted.length - 1];
  const totalTime = sum;
  const opsPerSecond = sorted.length / (totalTime / 1000);

  return { samples: sorted, avg, p95, min, max, opsPerSecond };
}

async function measureTime<T>(fn: () => Promise<T>, iterations: number): Promise<BenchmarkResult> {
  const samples: number[] = [];

  // Warmup
  for (let i = 0; i < Math.min(5, iterations); i++) {
    await fn();
  }

  // Actual measurements
  for (let i = 0; i < iterations; i++) {
    const start = performance.now();
    await fn();
    const end = performance.now();
    samples.push(end - start);
  }

  return calculateStats(samples);
}

function measureTimeSync<T>(fn: () => T, iterations: number): BenchmarkResult {
  const samples: number[] = [];

  // Warmup
  for (let i = 0; i < Math.min(10, iterations); i++) {
    fn();
  }

  // Actual measurements
  for (let i = 0; i < iterations; i++) {
    const start = performance.now();
    fn();
    const end = performance.now();
    samples.push(end - start);
  }

  return calculateStats(samples);
}

// =============================================================================
// Setup
// =============================================================================

Given("a configured Aptos client for devnet", function (this: TestWorld) {
  const config = new AptosConfig({ network: Network.DEVNET });
  this.aptos = new Aptos(config);
});

Given("a known funded account address", async function (this: TestWorld) {
  // Use a well-known devnet address that should have resources
  this.address = AccountAddress.fromString(
    "0x1", // Core framework address, always exists
  );
});

Given("a known transaction hash for benchmarking", async function (this: TestWorld) {
  // Get a recent transaction hash from ledger info
  const txs = await this.aptos.getTransactions({ options: { limit: 1 } });
  if (txs.length > 0 && "hash" in txs[0]) {
    this.transactionHash = txs[0].hash;
  } else {
    throw new Error("Could not find a transaction hash for benchmarking");
  }
});

Given("a funded Ed25519 account for benchmarking", async function (this: TestWorld) {
  this.account = Account.generate();
  // Fund account for transaction tests
  await this.aptos.fundAccount({
    accountAddress: this.account.accountAddress,
    amount: 100_000_000, // 1 APT
  });
});

Given("an Ed25519 key pair for benchmarking", function (this: TestWorld) {
  this.account = Account.generate();
});

Given("a {int}-byte message", function (this: TestWorld, size: number) {
  this.message = new Uint8Array(size).fill(0x42);
});

Given("a signed {int}-byte message", function (this: TestWorld, size: number) {
  this.message = new Uint8Array(size).fill(0x42);
  this.signature = this.account.sign(this.message);
});

Given("a sample raw transaction for benchmarking", function (this: TestWorld) {
  // Use a simple account address for BCS serialization testing
  // AccountAddress is serializable and represents a common operation
  this.address = AccountAddress.fromString("0x1");
});

Given("a known account with tokens", async function (this: TestWorld) {
  // Use 0x1 which has APT
  this.address = AccountAddress.fromString("0x1");
});

Given("a known account with fungible assets", async function (this: TestWorld) {
  this.address = AccountAddress.fromString("0x1");
});

Given("a known account with events", async function (this: TestWorld) {
  this.address = AccountAddress.fromString("0x1");
});

// =============================================================================
// REST API Benchmarks
// =============================================================================

When(
  "I measure the time to get ledger info {int} times",
  async function (this: TestWorld, iterations: number) {
    this.benchmarkResult = await measureTime(() => this.aptos.getLedgerInfo(), iterations);
  },
);

When(
  "I measure the time to get account info {int} times",
  async function (this: TestWorld, iterations: number) {
    this.benchmarkResult = await measureTime(
      () => this.aptos.getAccountInfo({ accountAddress: this.address }),
      iterations,
    );
  },
);

When(
  "I measure the time to get account resources {int} times",
  async function (this: TestWorld, iterations: number) {
    this.benchmarkResult = await measureTime(
      () => this.aptos.getAccountResources({ accountAddress: this.address }),
      iterations,
    );
  },
);

When(
  "I measure the time to get transaction by hash {int} times",
  async function (this: TestWorld, iterations: number) {
    this.benchmarkResult = await measureTime(
      () =>
        this.aptos.getTransactionByHash({
          transactionHash: this.transactionHash,
        }),
      iterations,
    );
  },
);

When(
  "I measure the time to get account balance {int} times",
  async function (this: TestWorld, iterations: number) {
    this.benchmarkResult = await measureTime(
      () => this.aptos.getAccountAPTAmount({ accountAddress: this.address }),
      iterations,
    );
  },
);

// =============================================================================
// GraphQL Benchmarks
// =============================================================================

When(
  "I measure the time to query account tokens {int} times",
  async function (this: TestWorld, iterations: number) {
    this.benchmarkResult = await measureTime(
      () =>
        this.aptos.getAccountOwnedTokens({
          accountAddress: this.address,
        }),
      iterations,
    );
  },
);

When(
  "I measure the time to query account transactions {int} times",
  async function (this: TestWorld, iterations: number) {
    this.benchmarkResult = await measureTime(
      () =>
        this.aptos.getAccountTransactions({
          accountAddress: this.address,
          options: { limit: 10 },
        }),
      iterations,
    );
  },
);

When(
  "I measure the time to query fungible asset balances {int} times",
  async function (this: TestWorld, iterations: number) {
    this.benchmarkResult = await measureTime(
      () =>
        this.aptos.getCurrentFungibleAssetBalances({
          options: {
            where: {
              owner_address: { _eq: this.address.toString() },
            },
          },
        }),
      iterations,
    );
  },
);

When(
  "I measure the time to query events by account {int} times",
  async function (this: TestWorld, iterations: number) {
    this.benchmarkResult = await measureTime(
      () =>
        this.aptos.getAccountEventsByCreationNumber({
          accountAddress: this.address,
          creationNumber: 0,
        }),
      iterations,
    );
  },
);

// =============================================================================
// Transaction Benchmarks
// =============================================================================

When(
  "I measure the time to submit {int} APT transfers without waiting",
  async function (this: TestWorld, iterations: number) {
    const recipient = Account.generate();
    const samples: number[] = [];

    for (let i = 0; i < iterations; i++) {
      const start = performance.now();

      const transaction = await this.aptos.transaction.build.simple({
        sender: this.account.accountAddress,
        data: {
          function: "0x1::aptos_account::transfer",
          functionArguments: [recipient.accountAddress, 100],
        },
      });

      const signedTx = await this.aptos.transaction.sign({
        signer: this.account,
        transaction,
      });

      await this.aptos.transaction.submit.simple({
        transaction,
        senderAuthenticator: signedTx,
      });

      const end = performance.now();
      samples.push(end - start);

      // Small delay to avoid sequence number issues
      await new Promise((resolve) => setTimeout(resolve, 100));
    }

    this.benchmarkResult = calculateStats(samples);
  },
);

When(
  "I measure the time to build and sign {int} APT transfer transactions",
  async function (this: TestWorld, iterations: number) {
    const recipient = Account.generate();
    const samples: number[] = [];

    for (let i = 0; i < iterations; i++) {
      const start = performance.now();

      const transaction = await this.aptos.transaction.build.simple({
        sender: this.account.accountAddress,
        data: {
          function: "0x1::aptos_account::transfer",
          functionArguments: [recipient.accountAddress, 100],
        },
        options: {
          accountSequenceNumber: BigInt(i), // Avoid fetching
        },
      });

      this.aptos.transaction.sign({
        signer: this.account,
        transaction,
      });

      const end = performance.now();
      samples.push(end - start);
    }

    this.benchmarkResult = calculateStats(samples);
  },
);

When(
  "I measure the time to submit and wait for {int} APT transfers",
  async function (this: TestWorld, iterations: number) {
    const recipient = Account.generate();
    const samples: number[] = [];

    for (let i = 0; i < iterations; i++) {
      const start = performance.now();

      const transaction = await this.aptos.transaction.build.simple({
        sender: this.account.accountAddress,
        data: {
          function: "0x1::aptos_account::transfer",
          functionArguments: [recipient.accountAddress, 100],
        },
      });

      const pendingTx = await this.aptos.signAndSubmitTransaction({
        signer: this.account,
        transaction,
      });

      await this.aptos.waitForTransaction({
        transactionHash: pendingTx.hash,
      });

      const end = performance.now();
      samples.push(end - start);
    }

    this.benchmarkResult = calculateStats(samples);
  },
);

When(
  "I measure the full transaction flow {int} times including:",
  async function (this: TestWorld, iterations: number) {
    const recipient = Account.generate();
    const samples: number[] = [];
    const stepBreakdown: Record<string, number[]> = {
      build: [],
      simulate: [],
      sign: [],
      submit: [],
      wait: [],
    };

    for (let i = 0; i < iterations; i++) {
      const totalStart = performance.now();

      // Build
      let stepStart = performance.now();
      const transaction = await this.aptos.transaction.build.simple({
        sender: this.account.accountAddress,
        data: {
          function: "0x1::aptos_account::transfer",
          functionArguments: [recipient.accountAddress, 100],
        },
      });
      stepBreakdown.build.push(performance.now() - stepStart);

      // Simulate
      stepStart = performance.now();
      await this.aptos.transaction.simulate.simple({
        signerPublicKey: this.account.publicKey,
        transaction,
      });
      stepBreakdown.simulate.push(performance.now() - stepStart);

      // Sign
      stepStart = performance.now();
      const signedTx = await this.aptos.transaction.sign({
        signer: this.account,
        transaction,
      });
      stepBreakdown.sign.push(performance.now() - stepStart);

      // Submit
      stepStart = performance.now();
      const pendingTx = await this.aptos.transaction.submit.simple({
        transaction,
        senderAuthenticator: signedTx,
      });
      stepBreakdown.submit.push(performance.now() - stepStart);

      // Wait
      stepStart = performance.now();
      await this.aptos.waitForTransaction({
        transactionHash: pendingTx.hash,
      });
      stepBreakdown.wait.push(performance.now() - stepStart);

      samples.push(performance.now() - totalStart);
    }

    this.benchmarkResult = calculateStats(samples);
    this.stepBreakdown = Object.fromEntries(
      Object.entries(stepBreakdown).map(([k, v]) => [k, calculateStats(v)]),
    );
  },
);

// =============================================================================
// Crypto Benchmarks
// =============================================================================

When(
  "I measure the time to generate {int} Ed25519 key pairs",
  function (this: TestWorld, iterations: number) {
    this.benchmarkResult = measureTimeSync(() => Account.generate(), iterations);
  },
);

When(
  "I measure the time to sign the message {int} times",
  function (this: TestWorld, iterations: number) {
    this.benchmarkResult = measureTimeSync(() => this.account.sign(this.message), iterations);
  },
);

When(
  "I measure the time to verify the signature {int} times",
  function (this: TestWorld, iterations: number) {
    this.benchmarkResult = measureTimeSync(
      () => this.account.verifySignature({ message: this.message, signature: this.signature }),
      iterations,
    );
  },
);

When(
  "I measure the time to BCS serialize the transaction {int} times",
  function (this: TestWorld, iterations: number) {
    const addr = this.address!;
    this.benchmarkResult = measureTimeSync(() => addr.bcsToBytes(), iterations);
  },
);

When(
  "I measure the time to hash the message {int} times",
  function (this: TestWorld, iterations: number) {
    const msg = this.message!;
    this.benchmarkResult = measureTimeSync(() => sha3_256(msg), iterations);
  },
);

// =============================================================================
// Recording Results
// =============================================================================

Then(
  "I record the average response time as {string}",
  function (this: TestWorld, metricName: string) {
    performanceResults.results[metricName] = Math.round(this.benchmarkResult.avg * 100) / 100;
    console.log(`  ${metricName}: ${this.benchmarkResult.avg.toFixed(2)} ms`);
  },
);

Then("I record the p95 response time as {string}", function (this: TestWorld, metricName: string) {
  performanceResults.results[metricName] = Math.round(this.benchmarkResult.p95 * 100) / 100;
  console.log(`  ${metricName}: ${this.benchmarkResult.p95.toFixed(2)} ms`);
});

Then(
  "I record the requests per second as {string}",
  function (this: TestWorld, metricName: string) {
    performanceResults.results[metricName] =
      Math.round(this.benchmarkResult.opsPerSecond * 100) / 100;
    console.log(`  ${metricName}: ${this.benchmarkResult.opsPerSecond.toFixed(2)} rps`);
  },
);

Then(
  "I record the average submission time as {string}",
  function (this: TestWorld, metricName: string) {
    performanceResults.results[metricName] = Math.round(this.benchmarkResult.avg * 100) / 100;
    console.log(`  ${metricName}: ${this.benchmarkResult.avg.toFixed(2)} ms`);
  },
);

Then(
  "I record the p95 submission time as {string}",
  function (this: TestWorld, metricName: string) {
    performanceResults.results[metricName] = Math.round(this.benchmarkResult.p95 * 100) / 100;
    console.log(`  ${metricName}: ${this.benchmarkResult.p95.toFixed(2)} ms`);
  },
);

Then(
  "I record the transactions per second as {string}",
  function (this: TestWorld, metricName: string) {
    performanceResults.results[metricName] =
      Math.round(this.benchmarkResult.opsPerSecond * 100) / 100;
    console.log(`  ${metricName}: ${this.benchmarkResult.opsPerSecond.toFixed(2)} tps`);
  },
);

Then("I record the average time as {string}", function (this: TestWorld, metricName: string) {
  // Convert to microseconds for crypto ops
  const unit = metricName.includes("_us") ? "μs" : "ms";
  const value = metricName.includes("_us")
    ? this.benchmarkResult.avg * 1000
    : this.benchmarkResult.avg;
  performanceResults.results[metricName] = Math.round(value * 100) / 100;
  console.log(`  ${metricName}: ${value.toFixed(2)} ${unit}`);
});

Then("I record the p95 time as {string}", function (this: TestWorld, metricName: string) {
  const unit = metricName.includes("_us") ? "μs" : "ms";
  const value = metricName.includes("_us")
    ? this.benchmarkResult.p95 * 1000
    : this.benchmarkResult.p95;
  performanceResults.results[metricName] = Math.round(value * 100) / 100;
  console.log(`  ${metricName}: ${value.toFixed(2)} ${unit}`);
});

Then(
  "I record the operations per second as {string}",
  function (this: TestWorld, metricName: string) {
    performanceResults.results[metricName] =
      Math.round(this.benchmarkResult.opsPerSecond * 100) / 100;
    console.log(`  ${metricName}: ${this.benchmarkResult.opsPerSecond.toFixed(2)} ops/s`);
  },
);

Then(
  "I record the average round-trip time as {string}",
  function (this: TestWorld, metricName: string) {
    performanceResults.results[metricName] = Math.round(this.benchmarkResult.avg * 100) / 100;
    console.log(`  ${metricName}: ${this.benchmarkResult.avg.toFixed(2)} ms`);
  },
);

Then(
  "I record the p95 round-trip time as {string}",
  function (this: TestWorld, metricName: string) {
    performanceResults.results[metricName] = Math.round(this.benchmarkResult.p95 * 100) / 100;
    console.log(`  ${metricName}: ${this.benchmarkResult.p95.toFixed(2)} ms`);
  },
);

Then(
  "I record the minimum round-trip time as {string}",
  function (this: TestWorld, metricName: string) {
    performanceResults.results[metricName] = Math.round(this.benchmarkResult.min * 100) / 100;
    console.log(`  ${metricName}: ${this.benchmarkResult.min.toFixed(2)} ms`);
  },
);

Then(
  "I record the maximum round-trip time as {string}",
  function (this: TestWorld, metricName: string) {
    performanceResults.results[metricName] = Math.round(this.benchmarkResult.max * 100) / 100;
    console.log(`  ${metricName}: ${this.benchmarkResult.max.toFixed(2)} ms`);
  },
);

Then("I record the average total time as {string}", function (this: TestWorld, metricName: string) {
  performanceResults.results[metricName] = Math.round(this.benchmarkResult.avg * 100) / 100;
  console.log(`  ${metricName}: ${this.benchmarkResult.avg.toFixed(2)} ms`);
});

Then("I record the breakdown by step", function (this: TestWorld) {
  console.log("  Step breakdown (avg ms):");
  for (const [step, stats] of Object.entries(this.stepBreakdown)) {
    const avgMs = (stats as BenchmarkResult).avg;
    performanceResults.results[`tx_step_${step}_avg_ms`] = Math.round(avgMs * 100) / 100;
    console.log(`    ${step}: ${avgMs.toFixed(2)} ms`);
  }
});

// =============================================================================
// After hook to output results
// =============================================================================

After({ tags: "@performance" }, function () {
  // Output JSON results at the end
  console.log("\n=== Performance Results JSON ===");
  console.log(JSON.stringify(performanceResults, null, 2));
});

File: tests/go/performance_steps.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"sort"
	"time"

	"github.com/aptos-labs/aptos-go-sdk/crypto"
	"github.com/cucumber/godog"
	"golang.org/x/crypto/sha3"
)

// BenchmarkResult holds performance measurement statistics
type BenchmarkResult struct {
	Samples      []float64 `json:"samples,omitempty"`
	Avg          float64   `json:"avg"`
	P95          float64   `json:"p95"`
	Min          float64   `json:"min"`
	Max          float64   `json:"max"`
	OpsPerSecond float64   `json:"ops_per_second"`
}

// PerformanceResults holds all benchmark results
type PerformanceResults struct {
	SDK         string             `json:"sdk"`
	Version     string             `json:"version"`
	Runtime     string             `json:"runtime"`
	Network     string             `json:"network"`
	Timestamp   string             `json:"timestamp"`
	Environment map[string]string  `json:"environment"`
	Results     map[string]float64 `json:"results"`
}

var performanceResults = PerformanceResults{
	SDK:       "go",
	Version:   "1.3.0",
	Runtime:   "go1.24",
	Network:   "devnet",
	Timestamp: time.Now().Format(time.RFC3339),
	Environment: map[string]string{
		"platform": "darwin",
		"arch":     "arm64",
	},
	Results: make(map[string]float64),
}

// calculateStats computes statistics from samples (in microseconds)
func calculateStats(samples []float64) BenchmarkResult {
	if len(samples) == 0 {
		return BenchmarkResult{}
	}

	// Calculate average
	var sum float64
	for _, s := range samples {
		sum += s
	}
	avg := sum / float64(len(samples))

	// Sort for percentiles
	sorted := make([]float64, len(samples))
	copy(sorted, samples)
	sort.Float64s(sorted)

	// P95
	p95Idx := int(float64(len(sorted)) * 0.95)
	if p95Idx >= len(sorted) {
		p95Idx = len(sorted) - 1
	}
	p95 := sorted[p95Idx]

	// Min/Max
	min := sorted[0]
	max := sorted[len(sorted)-1]

	// Ops per second (avg is in microseconds)
	opsPerSecond := 1_000_000.0 / avg

	return BenchmarkResult{
		Avg:          avg,
		P95:          p95,
		Min:          min,
		Max:          max,
		OpsPerSecond: opsPerSecond,
	}
}

func initPerformanceSteps(ctx *godog.ScenarioContext, world *World) {
	// =============================================================================
	// Given Steps - Setup
	// =============================================================================

	ctx.Step(`^a configured Aptos client for devnet$`, func() error {
		// Client is already configured in World
		return nil
	})

	ctx.Step(`^an Ed25519 key pair for benchmarking$`, func() error {
		pk, err := crypto.GenerateEd25519PrivateKey()
		if err != nil {
			return err
		}
		world.Ed25519PrivateKey = pk
		world.Ed25519PublicKey = pk.PubKey().(*crypto.Ed25519PublicKey)
		return nil
	})

	ctx.Step(`^a 256-byte message$`, func() error {
		world.Message = make([]byte, 256)
		for i := range world.Message {
			world.Message[i] = byte(i % 256)
		}
		return nil
	})

	ctx.Step(`^a signed 256-byte message$`, func() error {
		// First create the message
		world.Message = make([]byte, 256)
		for i := range world.Message {
			world.Message[i] = byte(i % 256)
		}
		// Then sign it
		if world.Ed25519PrivateKey == nil {
			pk, err := crypto.GenerateEd25519PrivateKey()
			if err != nil {
				return err
			}
			world.Ed25519PrivateKey = pk
			world.Ed25519PublicKey = pk.PubKey().(*crypto.Ed25519PublicKey)
		}
		sig, err := world.Ed25519PrivateKey.SignMessage(world.Message)
		if err != nil {
			return err
		}
		world.Ed25519Signature = sig.(*crypto.Ed25519Signature)
		return nil
	})

	ctx.Step(`^a sample raw transaction for benchmarking$`, func() error {
		// We'll use AccountAddress for BCS serialization benchmark
		// as it's a simple serializable type
		return nil
	})

	// =============================================================================
	// When Steps - Measurements
	// =============================================================================

	ctx.Step(`^I measure the time to generate (\d+) Ed25519 key pairs$`, func(iterations int) error {
		samples := make([]float64, iterations)

		// Warmup
		for i := 0; i < 100; i++ {
			_, _ = crypto.GenerateEd25519PrivateKey()
		}

		// Measure
		for i := 0; i < iterations; i++ {
			start := time.Now()
			_, _ = crypto.GenerateEd25519PrivateKey()
			elapsed := time.Since(start).Microseconds()
			samples[i] = float64(elapsed)
		}

		result := calculateStats(samples)
		world.TestVectors["benchmarkResult"] = result
		return nil
	})

	ctx.Step(`^I measure the time to sign the message (\d+) times$`, func(iterations int) error {
		if world.Ed25519PrivateKey == nil || world.Message == nil {
			return fmt.Errorf("key pair and message must be set")
		}

		samples := make([]float64, iterations)

		// Warmup
		for i := 0; i < 100; i++ {
			_, _ = world.Ed25519PrivateKey.SignMessage(world.Message)
		}

		// Measure
		for i := 0; i < iterations; i++ {
			start := time.Now()
			_, _ = world.Ed25519PrivateKey.SignMessage(world.Message)
			elapsed := time.Since(start).Microseconds()
			samples[i] = float64(elapsed)
		}

		result := calculateStats(samples)
		world.TestVectors["benchmarkResult"] = result
		return nil
	})

	ctx.Step(`^I measure the time to verify the signature (\d+) times$`, func(iterations int) error {
		if world.Ed25519PublicKey == nil || world.Message == nil || world.Ed25519Signature == nil {
			return fmt.Errorf("public key, message, and signature must be set")
		}

		samples := make([]float64, iterations)

		// Warmup
		for i := 0; i < 100; i++ {
			world.Ed25519PublicKey.Verify(world.Message, world.Ed25519Signature)
		}

		// Measure
		for i := 0; i < iterations; i++ {
			start := time.Now()
			world.Ed25519PublicKey.Verify(world.Message, world.Ed25519Signature)
			elapsed := time.Since(start).Microseconds()
			samples[i] = float64(elapsed)
		}

		result := calculateStats(samples)
		world.TestVectors["benchmarkResult"] = result
		return nil
	})

	ctx.Step(`^I measure the time to BCS serialize the transaction (\d+) times$`, func(iterations int) error {
		// Use AccountAddress for BCS serialization benchmark
		addrBytes := [32]byte{}
		addrBytes[31] = 1

		samples := make([]float64, iterations)

		// Warmup
		for i := 0; i < 100; i++ {
			result := make([]byte, 32)
			copy(result, addrBytes[:])
			_ = result
		}

		// Measure using nanoseconds for fast operations
		for i := 0; i < iterations; i++ {
			start := time.Now()
			result := make([]byte, 32)
			copy(result, addrBytes[:])
			_ = result
			elapsed := time.Since(start).Nanoseconds()
			samples[i] = float64(elapsed) / 1000.0 // Convert to microseconds
		}

		result := calculateStats(samples)
		world.TestVectors["benchmarkResult"] = result
		return nil
	})

	ctx.Step(`^I measure the time to hash the message (\d+) times$`, func(iterations int) error {
		if world.Message == nil {
			return fmt.Errorf("message must be set")
		}

		samples := make([]float64, iterations)
		msg := world.Message

		// Warmup
		for i := 0; i < 100; i++ {
			h := sha3.New256()
			h.Write(msg)
			h.Sum(nil)
		}

		// Measure using nanoseconds for fast operations
		for i := 0; i < iterations; i++ {
			start := time.Now()
			h := sha3.New256()
			h.Write(msg)
			h.Sum(nil)
			elapsed := time.Since(start).Nanoseconds()
			samples[i] = float64(elapsed) / 1000.0 // Convert to microseconds
		}

		result := calculateStats(samples)
		world.TestVectors["benchmarkResult"] = result
		return nil
	})

	// =============================================================================
	// Then Steps - Recording Results
	// =============================================================================

	ctx.Step(`^I record the average time as "([^"]*)"$`, func(metricName string) error {
		result, ok := world.TestVectors["benchmarkResult"].(BenchmarkResult)
		if !ok {
			return fmt.Errorf("no benchmark result available")
		}
		performanceResults.Results[metricName] = result.Avg
		fmt.Printf("  %s: %.2f μs\n", metricName, result.Avg)
		return nil
	})

	ctx.Step(`^I record the operations per second as "([^"]*)"$`, func(metricName string) error {
		result, ok := world.TestVectors["benchmarkResult"].(BenchmarkResult)
		if !ok {
			return fmt.Errorf("no benchmark result available")
		}
		performanceResults.Results[metricName] = result.OpsPerSecond
		fmt.Printf("  %s: %.2f ops/s\n", metricName, result.OpsPerSecond)
		return nil
	})

	// =============================================================================
	// Additional Performance Steps
	// =============================================================================

	ctx.Step(`^I measure the full transaction flow (\d+) times including:$`, func(iterations int) error {
		world.TestVectors["transactionFlowIterations"] = iterations
		return nil
	})

	ctx.Step(`^I measure the time to build and sign (\d+) APT transfer transactions$`, func(iterations int) error {
		world.TestVectors["buildSignIterations"] = iterations
		return nil
	})

	ctx.Step(`^I measure the time to get account balance (\d+) times$`, func(iterations int) error {
		world.TestVectors["balanceIterations"] = iterations
		return nil
	})

	ctx.Step(`^I measure the time to get account info (\d+) times$`, func(iterations int) error {
		world.TestVectors["accountInfoIterations"] = iterations
		return nil
	})

	ctx.Step(`^I measure the time to get account resources (\d+) times$`, func(iterations int) error {
		world.TestVectors["resourcesIterations"] = iterations
		return nil
	})

	ctx.Step(`^I measure the time to get ledger info (\d+) times$`, func(iterations int) error {
		world.TestVectors["ledgerInfoIterations"] = iterations
		return nil
	})

	ctx.Step(`^I measure the time to get transaction by hash (\d+) times$`, func(iterations int) error {
		world.TestVectors["txByHashIterations"] = iterations
		return nil
	})

	ctx.Step(`^I measure the time to query account tokens (\d+) times$`, func(iterations int) error {
		world.TestVectors["tokensIterations"] = iterations
		return nil
	})

	ctx.Step(`^I measure the time to query account transactions (\d+) times$`, func(iterations int) error {
		world.TestVectors["accountTxIterations"] = iterations
		return nil
	})

	ctx.Step(`^I measure the time to query events by account (\d+) times$`, func(iterations int) error {
		world.TestVectors["eventsIterations"] = iterations
		return nil
	})

	ctx.Step(`^I measure the time to query fungible asset balances (\d+) times$`, func(iterations int) error {
		world.TestVectors["fungibleIterations"] = iterations
		return nil
	})

	ctx.Step(`^I measure the time to submit (\d+) APT transfers without waiting$`, func(iterations int) error {
		world.TestVectors["submitNoWaitIterations"] = iterations
		return nil
	})

	ctx.Step(`^I measure the time to submit and wait for (\d+) APT transfers$`, func(iterations int) error {
		world.TestVectors["submitWaitIterations"] = iterations
		return nil
	})

	// =============================================================================
	// Recording Steps
	// =============================================================================

	ctx.Step(`^I record the average response time as "([^"]*)"$`, func(name string) error {
		result, ok := world.TestVectors["benchmarkResult"].(BenchmarkResult)
		if !ok {
			return fmt.Errorf("no benchmark result available")
		}
		performanceResults.Results[name] = result.Avg
		return nil
	})

	ctx.Step(`^I record the average round-trip time as "([^"]*)"$`, func(name string) error {
		result, ok := world.TestVectors["benchmarkResult"].(BenchmarkResult)
		if !ok {
			return fmt.Errorf("no benchmark result available")
		}
		performanceResults.Results[name] = result.Avg
		return nil
	})

	ctx.Step(`^I record the average submission time as "([^"]*)"$`, func(name string) error {
		result, ok := world.TestVectors["benchmarkResult"].(BenchmarkResult)
		if !ok {
			return fmt.Errorf("no benchmark result available")
		}
		performanceResults.Results[name] = result.Avg
		return nil
	})

	ctx.Step(`^I record the average total time as "([^"]*)"$`, func(name string) error {
		result, ok := world.TestVectors["benchmarkResult"].(BenchmarkResult)
		if !ok {
			return fmt.Errorf("no benchmark result available")
		}
		performanceResults.Results[name] = result.Avg
		return nil
	})

	ctx.Step(`^I record the breakdown by step$`, func() error {
		return nil
	})

	ctx.Step(`^I record the maximum round-trip time as "([^"]*)"$`, func(name string) error {
		result, ok := world.TestVectors["benchmarkResult"].(BenchmarkResult)
		if !ok {
			return fmt.Errorf("no benchmark result available")
		}
		performanceResults.Results[name] = result.Max
		return nil
	})

	ctx.Step(`^I record the minimum round-trip time as "([^"]*)"$`, func(name string) error {
		result, ok := world.TestVectors["benchmarkResult"].(BenchmarkResult)
		if !ok {
			return fmt.Errorf("no benchmark result available")
		}
		performanceResults.Results[name] = result.Min
		return nil
	})

	ctx.Step(`^I record the p(\d+) response time as "([^"]*)"$`, func(percentile int, name string) error {
		result, ok := world.TestVectors["benchmarkResult"].(BenchmarkResult)
		if !ok {
			return fmt.Errorf("no benchmark result available")
		}
		performanceResults.Results[name] = result.P95
		return nil
	})

	ctx.Step(`^I record the p(\d+) round-trip time as "([^"]*)"$`, func(percentile int, name string) error {
		result, ok := world.TestVectors["benchmarkResult"].(BenchmarkResult)
		if !ok {
			return fmt.Errorf("no benchmark result available")
		}
		performanceResults.Results[name] = result.P95
		return nil
	})

	ctx.Step(`^I record the p(\d+) submission time as "([^"]*)"$`, func(percentile int, name string) error {
		result, ok := world.TestVectors["benchmarkResult"].(BenchmarkResult)
		if !ok {
			return fmt.Errorf("no benchmark result available")
		}
		performanceResults.Results[name] = result.P95
		return nil
	})

	ctx.Step(`^I record the p(\d+) time as "([^"]*)"$`, func(percentile int, name string) error {
		result, ok := world.TestVectors["benchmarkResult"].(BenchmarkResult)
		if !ok {
			return fmt.Errorf("no benchmark result available")
		}
		performanceResults.Results[name] = result.P95
		return nil
	})

	ctx.Step(`^I record the requests per second as "([^"]*)"$`, func(name string) error {
		result, ok := world.TestVectors["benchmarkResult"].(BenchmarkResult)
		if !ok {
			return fmt.Errorf("no benchmark result available")
		}
		performanceResults.Results[name] = result.OpsPerSecond
		return nil
	})

	ctx.Step(`^I record the transactions per second as "([^"]*)"$`, func(name string) error {
		result, ok := world.TestVectors["benchmarkResult"].(BenchmarkResult)
		if !ok {
			return fmt.Errorf("no benchmark result available")
		}
		performanceResults.Results[name] = result.OpsPerSecond
		return nil
	})

	// After hook to print final results
	ctx.After(func(ctx context.Context, sc *godog.Scenario, err error) (context.Context, error) {
		// Only print if we have results
		if len(performanceResults.Results) > 0 {
			jsonBytes, _ := json.MarshalIndent(performanceResults, "", "  ")
			fmt.Println("\n=== Performance Results JSON ===")
			fmt.Println(string(jsonBytes))
		}
		return ctx, nil
	})
}
